internal/jobobject/iocp.go

```diff
@@ -0,0 +1,67 @@
+package jobobject
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	ioInitOnce sync.Once
+	initIOErr  error
+	// Global iocp handle that will be re-used for every job object
+	ioCompletionPort windows.Handle
+	// Mapping of job handle to channel that that job object owns to be able to receive
+	// notifications.
+	jobMap = make(map[uintptr]chan interface{})
+)
+
+// AllProcessesExited is a type representing a notification that every process in a job has exited.
+type AllProcessesExited struct {
+	Err error
+}
+
+// ioCompletionProcessor processes completed async IOs forever
+func pollIOCP(iocpHandle windows.Handle) {
+	var (
+		overlapped uintptr
+		code       uint32
+		key        uintptr
+	)
+
+	for {
+		err := winapi.GetQueuedCompletionStatus(iocpHandle, &code, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE)
+		if ch, ok := jobMap[key]; ok {
```

Comment:
If we receive a message for a key that is not in the map, I think we should log a warning. That could help us diagnose issues in the future.